File: 3/3.rb
@primes = [2]

def next_prime!
  i = @primes.last
  is_prime = false

  while !is_prime
    i += 1
    @primes.each do |prime|
      is_prime = i % prime != 0
      break unless is_prime
    end
  end
  @primes << i
  i
end

current_prime = 2

n = 600851475143
while n != 1
  while n % current_prime == 0
    n /= current_prime
  end

  current_prime = next_prime! unless n == 1
end
p current_prime

File: 3/3.ml
open Core.Std

let is_prime number primes =
  List.for_all primes ~f:(fun x -> (number mod x) <> 0)

let rec next_prime = function
  | []    -> assert false
  | (hd::_) as primes ->
      let rec search_prime primes current =
        if is_prime current primes
        then current
        else search_prime primes (current+2)
      in
        search_prime primes (hd+2)

let largest_prime_factor n =
  let rec search_largest_prime primes n current = match n with
    | 1 -> current
    | _ when (n mod current) = 0 -> search_largest_prime primes (n/current) current
    | _ -> let new_primes = current::primes in
             search_largest_prime new_primes n (next_prime new_primes)
  in
    search_largest_prime [2] n 3

let () = Printf.printf "%d\n" (largest_prime_factor 600851475143)
